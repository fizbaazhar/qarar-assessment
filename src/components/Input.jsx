import React from 'react';

/**
 * Input Component
 * Reusable input field with consistent styling and validation states
 * 
 * @param {string} label - Input label text
 * @param {string} type - Input type (text, email, password, etc.)
 * @param {string} placeholder - Placeholder text
 * @param {string} error - Error message to display
 * @param {boolean} disabled - Disables the input
 * @param {string} className - Additional CSS classes
 * @param {Object} props - All other input props
 */
const Input = ({ 
  label, 
  type = 'text', 
  placeholder, 
  error, 
  disabled = false, 
  className = '', 
  ...props 
}) => {
  return (
    <div className="mb-4">
      {label && (
        <label className="block text-gray-700 text-sm font-bold mb-2" htmlFor={props.id}>
          {label}
        </label>
      )}
      <input
        type={type}
        placeholder={placeholder}
        className={`
          shadow appearance-none border rounded-lg w-full py-3 px-4 text-gray-700 leading-tight 
          focus:outline-none focus:ring-2 focus:ring-primary focus:border-transparent transition-all
          ${error ? 'border-red-300' : 'border-gray-200'}
          ${disabled ? 'opacity-50 cursor-not-allowed' : ''}
          ${className}
        `}
        disabled={disabled}
        {...props}
      />
      {error && (
        <p className="text-red-500 text-xs mt-1">{error}</p>
      )}
    </div>
  );
};

export default Input; 