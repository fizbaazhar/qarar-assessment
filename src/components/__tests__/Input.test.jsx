import { describe, it, expect, vi } from 'vitest'
import { render, screen, fireEvent, waitFor } from '../../test/test-utils'
import userEvent from '@testing-library/user-event'
import Input from '../Input'

describe('Input Component', () => {
  // Test basic rendering
  it('renders input with label', () => {
    render(<Input label="Email" id="email" />)
    expect(screen.getByLabelText('Email')).toBeInTheDocument()
  })

  it('renders input with placeholder', () => {
    render(<Input label="Email" placeholder="Enter your email" />)
    expect(screen.getByPlaceholderText('Enter your email')).toBeInTheDocument()
  })

  // Test different input types
  it('renders email input type correctly', () => {
    render(<Input label="Email" type="email" />)
    const input = screen.getByRole('textbox')
    expect(input).toHaveAttribute('type', 'email')
  })

  it('renders password input type correctly', () => {
    render(<Input label="Password" type="password" />)
    const input = screen.getByDisplayValue('')
    expect(input).toHaveAttribute('type', 'password')
  })

  it('renders number input type correctly', () => {
    render(<Input label="Age" type="number" />)
    const input = screen.getByRole('spinbutton')
    expect(input).toHaveAttribute('type', 'number')
  })

  // Test value and onChange
  it('displays the provided value', () => {
    render(<Input label="Email" value="test@example.com" onChange={() => {}} />)
    expect(screen.getByDisplayValue('test@example.com')).toBeInTheDocument()
  })

  it('calls onChange when user types', async () => {
    const user = userEvent.setup()
    const handleChange = vi.fn()
    render(<Input label="Email" onChange={handleChange} />)
    
    const input = screen.getByRole('textbox')
    await user.type(input, 'test')
    
    expect(handleChange).toHaveBeenCalledTimes(4) // Once for each character
  })

  // Test error state
  it('displays error message when provided', () => {
    render(<Input label="Email" error="Email is required" />)
    expect(screen.getByText('Email is required')).toBeInTheDocument()
    expect(screen.getByText('Email is required')).toHaveClass('text-red-500')
  })

  it('applies error styling to input when error is present', () => {
    render(<Input label="Email" error="Email is required" />)
    const input = screen.getByRole('textbox')
    expect(input).toHaveClass('border-red-300')
  })

  // Test disabled state
  it('disables input when disabled prop is true', () => {
    render(<Input label="Email" disabled={true} />)
    const input = screen.getByRole('textbox')
    expect(input).toBeDisabled()
  })

  // Test required attribute
  it('applies required attribute when required prop is true', () => {
    render(<Input label="Email" required={true} />)
    const input = screen.getByRole('textbox')
    expect(input).toBeRequired()
  })

  // Test name attribute
  it('applies name attribute correctly', () => {
    render(<Input label="Email" name="email" />)
    const input = screen.getByRole('textbox')
    expect(input).toHaveAttribute('name', 'email')
  })

  // Test id and htmlFor connection
  it('connects label to input via id and htmlFor', () => {
    render(<Input label="Email" id="email" />)
    const input = screen.getByRole('textbox')
    const label = screen.getByText('Email')
    
    expect(input).toHaveAttribute('id', 'email')
    expect(label).toHaveAttribute('for', 'email')
  })

  // Test focus states
  it('applies focus styles when input is focused', async () => {
    const user = userEvent.setup()
    render(<Input label="Email" />)
    
    const input = screen.getByRole('textbox')
    await user.click(input)
    
    expect(input).toHaveFocus()
  })

  // Test snapshot
  it('matches snapshot for basic input', () => {
    const { container } = render(<Input label="Email" placeholder="Enter email" />)
    expect(container.firstChild).toMatchSnapshot()
  })

  it('matches snapshot for input with error', () => {
    const { container } = render(
      <Input label="Email" error="Email is required" />
    )
    expect(container.firstChild).toMatchSnapshot()
  })
}) 