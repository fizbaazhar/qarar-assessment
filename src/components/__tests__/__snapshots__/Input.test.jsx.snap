// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Input Component > matches snapshot for basic input 1`] = `
<div
  class="mb-4"
>
  <label
    class="block text-gray-700 text-xs sm:text-sm font-bold mb-2"
  >
    Email
  </label>
  <input
    class="
          shadow appearance-none border rounded-lg w-full py-2 sm:py-3 px-3 sm:px-4 text-xs sm:text-sm text-gray-700 leading-tight 
          focus:outline-none focus:ring-2 focus:ring-primary focus:border-transparent transition-all
          border-gray-200
          
          
        "
    placeholder="Enter email"
    type="text"
  />
</div>
`;

exports[`Input Component > matches snapshot for input with error 1`] = `
<div
  class="mb-4"
>
  <label
    class="block text-gray-700 text-xs sm:text-sm font-bold mb-2"
  >
    Email
  </label>
  <input
    class="
          shadow appearance-none border rounded-lg w-full py-2 sm:py-3 px-3 sm:px-4 text-xs sm:text-sm text-gray-700 leading-tight 
          focus:outline-none focus:ring-2 focus:ring-primary focus:border-transparent transition-all
          border-red-300
          
          
        "
    type="text"
  />
  <p
    class="text-red-500 text-xs mt-1"
  >
    Email is required
  </p>
</div>
`;
